### ISSUE TITLE
LJSPECCH synthesis quality has some issue

### ISSUE THREAD

**Issue by nvadigauvce** [espnet/espnet#1878]
Hi All,

I have trained LJSPEECH with tts1 recipe with default configuration (train_pytorch_tacotron2.v3.yaml) for 400 epochs both char and phone based model with 8 GPU, I used parallel wavegan for synthesis. But I found some issues, which are listed below:

1)  Synthesis quality is ok, but Amplitude levels of synthesis is lower compared to original speech. Have you seen this issue before? Or do we need to do any kind of normalization before training ?

2) I am also facing this issue of Tacotron2 model sometimes causes the deletion of the words or fails to stop the generation for longer and unseen texts, Is there any solution to this problem ?

3) Is it possible to know what is the maximum duration we can generate with Tacotron 2 model, because when I tried to generate audio size more than 2 min, tacotron 2 starts blabbering.

4) How important is the mel-range of 80-7600 in default configuration?, I have tried with both 0-8000 and 125-7600, I didnt found much difference.

**Comment by kan-bayashi** [comment]
1. We perform mean-var normalization for feature sequence but do not perform the normalization for waveform. If it is the problem, it is worthwhile to tune the global gain in parallel wavegan config.
https://github.com/kan-bayashi/ParallelWaveGAN/blob/cc1753289fcee3d3a230d18fa2e71d77abaf92df/egs/ljspeech/voc1/conf/parallel_wavegan.v1.yaml#L18

2. One of the solutions is to use the attention constraint in decoding. In my case, it works well.
https://github.com/espnet/espnet/blob/75d8acea455c54346ec879a38049046f7be827d6/egs/ljspeech/tts1/conf/decode.yaml#L5

3. I'm not sure since I did not try such a long sentence. But if we use the above attention constraint, it works well for the long sentence. 

4. We have some discussion. https://github.com/espnet/espnet/issues/1192
I did not compare with subjective evaluation but it is expected that the limit of Mel range is effective to avoid the oversmoothing problem.

**Comment by nvadigauvce** [comment]
Thanks @kan-bayashi for the quick reply. 
1. I will try to increase global gain and see if it helps in improving quality. Do you think if we use phone based model, which consist of more phonemes compared to grapheme respresenations, do we need more number of epochs to get good model or default settings are sufficient ?

2 & 3. use-att-constraint: True, helped at least with my limited testing for to solve some of the missing words.

4. Limiting the mel-range will help in faster convergence ? Also, in case of multi-speaker scenarios, how we can set the mel-range to avoid the pitch range of Male speakers ?

**Comment by kan-bayashi** [comment]
> 1. Do we need more number of epochs to get good model or default settings are sufficient ?

The quality of the generative model will be improved even if the model is overfitted. Therefore, it is worthwhile to train more epochs.

> 4. Limiting the mel-range will help in faster convergence ? Also, in case of multi-speaker scenarios, how we can set the mel-range to avoid the pitch range of Male speakers?

I've not checked the convergence speed. If there are speakers whose f0 is very low, it is better to use lower bound but I have not so much experience of the multi-speaker condition. Sorry.

**Comment by nvadigauvce** [comment]
> The quality of the generative model will be improved even if the model is overfitted. Therefore, it is worthwhile to train more epochs.

@kan-bayashi : Thanks for clarification, will try to run for some more epochs. Some more question are:
1) how model.loss.best is selected in ESPNET TTS model? Is it just based validation/main/loss (L1+MSE+BCE+Attention) or we using any other criteria?

2) What are the advantages of using L1 and BCE loss, when compared Nvidia code (where it uses only MSE). In your earlier post it was mentioned that "we use both l1 and mse losses to make the attention learning faster". 

a) So that means we need less number of epochs to train if we L1 loss?

b) what is the advantage of using BCE loss?

**Comment by kan-bayashi** [comment]
1. Yes, loss best is based on /main/validation/loss.

2. a) Not sure but in terms of attention L1 loss leads diagonal attention much faster.

b) BCE loss is for stop token prediction. Not for mel-spectrogram prediction.

**Comment by dathudeptrai** [comment]
hi @kan-bayashi , @nvadigauvce. The problem about Amplitude levels of synthesis is lower compared to original speech cause by gan training manner on PWG. Pls look https://github.com/kan-bayashi/ParallelWaveGAN/issues/131, it's fixed now if u use new manner. BTW it maybe cause by Tacotron-2 on non-teacher forcing condition, pls check if u use schedule teacher forcing on training tacotron-2.

**Comment by nvadigauvce** [comment]
@kan-bayashi @dathudeptrai thanks for the detailed answers.

@dathudeptrai 1) I was using old PWG code, so I will use latest PWG where re-computation of output of Generator is done before passing into discriminator.

2) I think by default in ESPNET, training of Tacotron-2 is done in teacher forcing condition. So do you feel we have to stop teacher forcing training after some epochs to get better quality ?

**Comment by dathudeptrai** [comment]
@kan-bayashi , @nvadigauvce hi, i just rethinking about tacotron-2 implementation and did some experiments, this is my conclusion about teacher forcing on tacotron-2.


1. if we Training tacotron with teacher forcing + 0.5 dropout rate for prenet then on an inference steps, we disable dropout ->>> alignment exploding. Why ? because in the inference step, we don't have real prev_mel so the decoder cann't adapt with real prev_mel.
2. Even apply 0.8-0.9 dropout rate for prenet, the model can convergence. The model just really need encoder hidden state and query to predict mel.
3. Apply norm for prenet can improve quality. Why ?. That is because when we norm prenet, we reduce internal covariate shift then a noise from real prev mel will be less impact on the decoder for the time being.
4. Training teacher forcing with 0.1 for dropout rate not as good as 0.5 (dropout apply for both training and inference)

conclusion: The bigger the dropout, the less the effect of prev mel on the current decoder, resulting in the noise of prev mel during actual inference will have little effect on the model results. (ofc we need apply dropout on inference step also =))))

**Comment by kan-bayashi** [comment]
@dathudeptrai Thank you for sharing great insight! So if we apply dropout for both training and inference, scheduled sampling is not so important, right?

**Comment by dathudeptrai** [comment]
@kan-bayashi absolutely yes, the model with 0.5 dropout is better than 0.1 dropout significantly.

**Comment by nvadigauvce** [comment]
@dathudeptrai As of now we are already using dropout of 0.5 both in training and evaluation, so do you suggest to further increase the dropout to 0.8-0.9 to improve the quality ?

**Comment by dathudeptrai** [comment]
If ljspeech, 0.5 is ok. Just try and see :))). On my multispeaker dataset dropout 0.7 is the best
